docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,423 @@
+Custom Attestation Data Formats for Open Enclave V3 Updates
+================
+
+As custom attestation plugins are developed and features are requested, new APIs were added to:
+
+1. Query built-in SGX attesters and verifiers to be registered.
+2. Select a format of the evidence that the registered attesters can generate.
+3. Query evidence parameters to be used as optional inputs when generating evidences.
+
+The following API was affected:
+
+1. Built-in SGX attesters and verifiers
+
+This document describes the added and affected APIs as well as keeping the existing support for version 2 of the API. This document assumes the user is familiar with the SGX attestation and [custom attestation](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/CustomAttestation.md) with plugins.
+
+Motivation
+----------
+
+New APIs were added to help customers who would like to extend Open Enclave's current attestation structure with extra information while keep TEE-specific information within attester and verifier plugins.
+
+1. The application can query built-in SGX attesters and verifiers to be registered. This allows the plugin library to return attesters and verifiers based on configuration.
+
+```C
+   // Old
+   oe_register_attester(oe_sgx_attesterX(), NULL, 0);
+   oe_register_attester(oe_sgx_attesterY(), NULL, 0);
+   oe_register_attester(oe_sgx_attesterZ(), NULL, 0);
+
+   // New
+   oe_sgx_get_attesters(&attesters, &attesters_length);
+   for (n = 0; n < attesters_length; n++)
+      oe_register_attester(&attesters[n], NULL, 0);
+```
+
+2. The attester application can rely on the SDK to select a matching evidence format.
+
+```C
+   // Old
+   const oe_uuid_t attester_formats[3] = {{ATTESTERX_UUID}, {ATTESTERY_UUID}, {ATTESTERZ_UUID}};
+   selected_format = NULL;
+   for (n = 0; n < requested_formats_length && !selected_format; n++)
+   {
+       for (m = 0; m < countof(attester_formats) && !selected_format; m++)
+       {
+           if (requested_formats[n] == attester_formats[m])
+               selected_format = &attester_formats[m];
+       }
+   }
+
+   // New
+   oe_select_attester_format(
+       requested_formats,
+       requested_formats_length,
+       &selected_format);
+```
+
+3. The application gets optional evidence parameters from verifiers and stays TEE-agnostic when generating evidences.
+
+```C
+   // Old
+   uint8_t opt_paramsX[] = { OPT_PARAMS_X };
+   uint8_t opt_paramsY[] = { OPT_PARAMS_Y };
+   if (format == ATTESTERX_UUID)
+   {
+      oe_get_evidence(&format, 0, NULL, 0, opt_paramsX, sizeof(opt_paramsX), &evidence, &evidence_size, NULL, NULL);
+   }
+   else if (format == ATTESTERY_UUID)
+   {
+       oe_get_evidence(&format, 0, NULL, 0, opt_paramsY, sizeof(opt_paramsY), &evidence, &evidence_size, NULL, NULL);
+   }
+
+   // New
+   uint8_t* opt_params;
+   size_t opt_params_size;
+   oe_get_evidence_parameters(&format, &opt_params, &opt_params_size);
```

Comment:
I am not clear why have the user application get the evidence parameters at all given that the plug-in itself has these parameters already?

Reply:
The parameters go from verifier side to attester side. The information is from the plug-ins but it needs to be transmitted by the applications. Hence, the public API.

---
In reply to: [392772865](https://github.com/openenclave/openenclave/pull/2621#discussion_r392772865) [](ancestors = 392772865)

Reply:
What does "The parameters go from verifier side to attester side" mean?   I believe there should be no such requirement in a protocol in general.

Reply:
Some evidence generation needs information from verifier side. For example, SGX (local) report generation needs 'target info' from verifier so the evidence can only be verified by the requested verifier.

We are not defining protocol here but just want to make the API capable of handling most of the known attestation systems.

---
In reply to: [393249417](https://github.com/openenclave/openenclave/pull/2621#discussion_r393249417) [](ancestors = 393249417)

Reply:
> Some evidence generation needs information from verifier side. For example, SGX (local) report generation needs 'target info' from verifier so the evidence can only be verified by the requested verifier.
> We are not defining protocol here but just want to make the API capable of handling most of the known attestation systems.

That's a reasonable answer.  Please update the doc to explain that.

Reply:
Updated motivation 3 and reworded user experience.

---
In reply to: [394651508](https://github.com/openenclave/openenclave/pull/2621#discussion_r394651508) [](ancestors = 394651508)